jupyterhub_moss/spawner.py

```diff
@@ -351,70 +342,103 @@ def options_from_form(self, formdata: Dict[str, List[str]]) -> Dict[str, str]:
 
         self.__validate_options(options)
 
-        partition = options["partition"]
+        partition_environments = tuple(
+            self.partitions[options["partition"]]["jupyter_environments"].values()
+        )
+        if "environment_path" not in options:
+            # Set path to use from first environment for the current partition
+            options["environment_path"] = partition_environments[0]["path"]
+
+        corresponding_default_env = find(
+            lambda env: env["path"] == options["environment_path"],
+            partition_environments,
+        )
+        # Singularity images are never added to PATH,
+        # custom envs are always added to PATH, defaults ones only if add_to_path is True
+        if not options["environment_path"].endswith(".sif") and (
+            corresponding_default_env is None
+            or corresponding_default_env["add_to_path"]
+        ):
+            options["prologue"] = f"export PATH={options['environment_path']}:$PATH"
+
+        return options
+
+    def __check_user_options(self, partition_info):
+        """Check if requested resources are valid for the given partition info.
 
+        Raises AssertionError if request does not match available resources.
+        """
+        if "runtime" in self.user_options:
+            runtime = parse_timelimit(self.user_options["runtime"])
+            assert (
+                runtime.total_seconds() <= partition_info["max_runtime"]
+            ), "Requested runtime exceeds partition time limit"
+
+        if self.user_options.get("nprocs", -1) > partition_info["max_nprocs"]:
```

Comment:
I feel your previous syntax `if "nprocs" in self.user_options and self.user_options["nprocs"] > partition_info["max_nprocs"]` read better